### ISSUE TITLE
Suggestion: implementing optional wildcard permissions

### ISSUE THREAD

**Issue by stefanriehl** [spatie/laravel-permission#1373]
I have a suggestion for the implementation of wildcard permissions based on [Apache Shiro](https://shiro.apache.org/permissions.html). Would that be interesting or is that not an option? Idea is to have the standard implementation like it is, but optionally set in config use_wildcard_permissions to true (default is false). If so, wildcard permissions will be used.

If that is still an option, I would do a pull request.

**Comment by drbyte** [comment]
@stefanriehl thanks for the suggestion!

I'm interested in your thoughts about the following:

- Laravel tends to lean towards dot-notation. Shiro tends towards colon-notation (granted it's got a purpose). Would you still want to use Shiro-style delimiters?
- How would implementing Shiro's approach require db storage changes? Do you envision that assigning multiple (ie: Shiro's comma-delimited) permissions would just insert multiple relational records behind-the-scenes?
- How do you envision linking Shiro's 3rd tier assignments to specific models?
- What do you think of the code in #440 ? What limitations does it have that Shiro's approach would solve?

Those are the first few questions that come to mind.

**Comment by stefanriehl** [comment]
- Would you still want to use Shiro-style delimiters?

This is actually configurable. It's also possible to use dot-notation. Examples would be:

```php
$permission = 'user.edit,view,delete';
$permission = 'user,post.*'; // same like $permission = 'user,post';
$permission = 'post.edit.456';
$permission = 'comment.*.762';
$permission = '*.view';
```

- How would implementing Shiro's approach require db storage changes? Do you envision that assigning multiple (ie: Shiro's comma-delimited) permissions would just insert multiple relational records behind-the-scenes?

The implementation would not require any changes to the database. The permission would be stored like the existing one's.

- How do you envision linking Shiro's 3rd tier assignments to specific models?

Hmm, IMO that should be part of the application logic. Let's say a user can only edit and delete it's own blog post: I would store a direct permission to the user. 

```php
$user->givePermissionTo('post.view,edit,delete.42');

// in blade
@can('post.delete.' . $post->id)
// ...
@endcan
```

- What do you think of the code in #440 ? What limitations does it have that Shiro's approach would solve?

I think, Shiro's approach is a bit more flexible using wildcards (for each part) and multiple values per part.

**Comment by drbyte** [comment]
Fair points.

I guess, like #440, this doesn't change db schema because it still represents all the expressiveness of meaning in a mere "string". It's down to the middleware to parse out the Shiro-style meanings at runtime.

I think for something like "an author can edit and delete their own posts" I'd probably not even generate a permission record, since their identity is already linked to the post. Less records to parse that way.

While possible to do already today, I've observed resistance to the idea of creating permissions which name specific object IDs to which they apply. But architecturally that's still the right way unless one extends things to add additional pivots and code to support those pivots, but that's all tangental to this discussion.

I've long envisioned including code much like what's in #440 in a future major release. So I'm particularly interested in finding the right intersection between the two approaches.
Maybe it's just a matter of configuration. Or maybe it's just a choice of whether or not to use helper methods for making assignments, as the middleware could be (nearly, if not fully) the same.

**Comment by stefanriehl** [comment]
I would leave the middleware as it is and not change it. The only change would be in the HasPermissions Trait. Then if it would be optional:

```php
// config/permission.php
'use_wildcard_permissions' => true
```

This would be triggered in `hasPermissionTo` method. I do not know if this is an option for you to have two possibilities to write permission strings.

Sometimes in our applications we don't know yet if a user is allowed to change his own post (as an example) or whatever resource. This is partially configurable, for example for an administrator.

The 3rd part of the string does not necessarily have to be an object ID. Shiro calls it a target. But in most cases it will be an ID, you're right.

**Comment by drbyte** [comment]
> Sometimes in our applications we don't know yet if a user is allowed to change his own post (as an example) or whatever resource. This is partially configurable, for example for an administrator.

Agreed. But as you say, that's application-layer logic. A feature-switch could be checked in the Policy that implements the authorization for that action, either in-conjunction-with, or independent-of any assigned permissions.

Admittedly this is more of a separate topic, but I see many people coupling the two issues at the "name of the permission" level, rather than in application logic. Hence the mention of it.

**Comment by drbyte** [comment]
@benyanke Thoughts?

**Comment by drbyte** [comment]
@stefanriehl wrote:
> Shiro's approach is a bit more flexible using wildcards (for each part) and multiple values per part.

So, just to be sure I'm seeing it how you're seeing it:
- it's offering the use of `*` in each part, and the comma-delimited multiple values in each part. 
- and that's what we'd need to accommodate in the parsing activities of the `hasPermissionTo` method.

> set in config use_wildcard_permissions to true (default is false)

Naming is hard. I'm a little concerned that many people will interpret the word `wildcard` with wildly different meanings. I welcome discussion for/against, and alternatives.

> I do not know if this is an option for you to have two possibilities to write permission strings.

What do you mean by "two possibilities to write permission strings"?

**Comment by drbyte** [comment]
@eduardoarandah Thoughts?

**Comment by stefanriehl** [comment]
> it's offering the use of * in each part, and the comma-delimited multiple values in each part.

Yes, but to be honest: there can be as many parts and subparts as you want. There is no limitation on exactly 3 parts. And there is no fixed logic in "{resource}.{action}.{objectId}". Your application decides that! But this is a common real life structure. And by using * wildcards in each part you will have a powerful tool.

Structure of the string

```php
// {part 1} could be * OR {subpart1},{subpart2},{subpart n}
// {part 2} could be * OR {subpart1},{subpart2},{subpart n}
// {part 3} could be * OR {subpart1},{subpart2},{subpart n}
// {part n} could be * OR {subpart1},{subpart2},{subpart n}
```

To make it more clear with some examples:

```php

// grants all permissions
$permission = '*';

// grants all actions on posts
$permission = 'posts.*';
// same as
$permission = 'posts';

// grants all actions on posts, projects, users and orders
$permission = 'posts,projects,users,orders.*';
// same as
$permission = 'posts,projects,users,orders';


// grants actions create, edit, view on all resources
$permission = '*.create,edit,view';

// even this is possible (if it makes any sense)
$permission = 'admin.users.*.delete.whatever.*.want';
// or
$permission = 'users,permissions,roles.create,edit,view,delete,list';

// you could use Class names as well
$permission = 'User::class.*.admin,editor';
```

It is important to mention, that the permission to be checked does not have to be stored in the permissions table. I don't know if that was clear.

```php
// example
// permissions in the database
$permissions = [
    '*', '*.view', 'posts', 'posts.create,view', 'posts.*.sports'
];

// permission to be checked 
// posts.view.sports is not in the $permissions array
$user->can('posts.view.sports')
// all $permissions are valid
```

> set in config use_wildcard_permissions to true (default is false)
> I do not know if this is an option for you to have two possibilities to write permission strings.

My first thought was to offer another implementation in addition to the existing one via settings in the config file, but by now it's clear to me that we don't need that at all. We could simply enable or disable the feature via config file (with a better name;-).

```php
/*
* ...
* @param string|int|\Spatie\Permission\Contracts\Permission $permission
* ...
*/
public function hasPermissionTo($permission, $guardName = null): bool
{
    // ...
    // if existing implementation returns false
    // and $permission is string
    // and wildcard permissions are enabled in config
    // check if user has permission via wildcard
}
```

**Comment by eduardoarandah** [comment]
@drbyte 

I think the star notation could be a little bit tricky to implement, it may require `like %%` statements in database and that doesn't sound efficient. 

Also, permissions are cached as a key-value aren't they? 

In this implementation I simply split permissions by the dot and check one by one until a match is found:

https://github.com/eduardoarandah/spatie-laravel-permission-permission-middleware

It doesn't have side-effects with regular permissions, unless they have a "." so it lets the user grow organically in their apps. 

When the app is a baby.. user will have "user" permission, then it can grow to "user.modify" 
then "user.modify.delete" 

In my humble opinion, more than that could be overengineering

**Comment by drbyte** [comment]
@eduardoarandah Yes, the cascading lookup you implemented only limits things by the `.` character, and this wildcarding would also add the `,` and `*` characters. It may also add some curious complexity if it were to support model-class-names (using `::`) as well. If architected strategically we can permit people to avoid the unnecessary lookups based on how they implement it in their own apps.

**Comment by drbyte** [comment]
@stefanriehl, Yes, I'd entertain a tested PR for this. Might be good to assess performance as well.

You're probably already considering this, but I'll mention it so you know what I'm thinking: best to balance out the implementation with some public methods that are easily overridden to avoid time-consuming lookups that people may not want, or they could use to implement even more varied lookups based on localized app needs.
That said, the package traits already have a lot of methods, so let's try not to clutter things either. :)

**Comment by stefanriehl** [comment]
@drbyte I am on vacation until Monday and will try to make a Pull Request afterwards.

@eduardoarandah I won‘t do a database lookup with % like or something like that. The db won‘t be changed and no special queries at all.

Let‘s see what I can do regarding performance. But first tests with a check against ~1000 user permissions looked promising.

Btw. Thanks guys for your time :)

**Comment by stefanriehl** [comment]
ok, I could do a [PR](https://github.com/spatie/laravel-permission/pull/1381) sooner as I thought with tests and documentation.

**Comment by drbyte** [comment]
Closed by #1381, released in https://github.com/spatie/laravel-permission/releases/tag/3.9.0

**Comment by doraemonxxx** [comment]
Hello all, Is there any way to change the delimiter of dot `.` into a colon `:`  ?
